contents/security/sqlInjection.md

```diff
@@ -13,86 +13,132 @@
+**However, some websites may not check the syntax of user input rigorously, and therefore a malicious user can inject SQL query via the user input.** 
+
+In the example above, the website directly substitutes the information given by the user without any validation. In this case, an attacker can supply some malicious SQL code in the user input such that it changes the nature of the SQL statement executed.
+
+ For example, the attacker can add more parameters to the query:
 
 ```sql
-Username: foo
-Password: bar’ OR ‘’=’
+Username: Admin
+Password: foo’ OR ‘1’=‘1
 ```
-
 The SQL command string built from this input would be as follows:
 
 ```sql
-SELECT * FROM Users WHERE Username = ‘foo’ 
-AND Password = ‘bar’ OR ‘’=’’
+SELECT * FROM Users WHERE Username = ‘Admin’ 
+AND Password = ‘foo’ OR ‘1’=‘1’
 ```
 
 In SQL, `AND` operations are checked before `OR` operations.
 This query will check the database for entries where:
 
 ```
-Username = foo AND Password = bar
-OR
-‘’=’’
+(Username = Admin AND Password = foo) OR (‘1’=‘1’)
 ```
 
-This query will always return true, as `‘’=’’` is always true.
+This `where` clause will always return true, as `‘1’=‘1’` is always true.
 As such, the query can be simplified to this:
 
 ```sql
 SELECT * FROM Users
 ```
 
-This will return all the rows from the `Users` table in the database, regardless of username or password entered. If used in user authentication, the user will likely be able to gain access to the system.
+This will return all the rows from the `Users` table in the database, regardless of username or password entered. 
+
+The above technique of injecting malicious SQL code via user input is called SQL injection. If used in user authentication, the attacker is able to gain access to anyone's account. Moreover, this attacker can also modify sensitive information if the account owner has the privilege (e.g. a lecturer who can modify students' marks).
+
+
+**In addition to adding extra parameters to compromise the authentication, a malicious user may even add custom queries to view, modify the records in database, or even delete the whole database.**
+
+An SQL query ends with a semicolon ";". In the previous section the malicious user terminates one parameter by single quote "'", and add more parameters behind it. Now, he can also terminate the query by semicolon, and adds another query at the back:
+
+```sql
+Username: foo
+Password: bar’; DROP TABLE Users;
+```
+
+```sql
+SELECT * FROM Users WHERE Username = ‘foo’ AND Password = ‘bar’; 
+DROP TABLE Users;
+```
+
+Then the database will execute these two queries, and delete all user information. Then other users cannot access this website. In addition to the `DROP` query, the attacker may also inject `SELECT` and `INSERT` queries, which can either read sensitive data from database or add data to it. 
```

Comment:
> Then the database will execute these two queries, and delete all user information. 

->

>When the database executes these two queries, it will delete all user information.